presto-main/src/main/java/com/facebook/presto/type/Re2JRegexp.java

```diff
@@ -0,0 +1,160 @@
+        }
+        else {
+            re2jPatternWithoutDotStartPrefixSuffix = re2jPattern;
+        }
+    }
+
+    public boolean matches(Slice source)
+    {
+        return re2jPatternWithoutDotStartPrefixSuffix.find(source);
+    }
+
+    public Slice replace(Slice source, Slice replacement)
+    {
+        Matcher matcher = re2jPattern.matcher(source);
+        try {
+            return matcher.replaceAll(replacement);
+        }
+        catch (IndexOutOfBoundsException | IllegalArgumentException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacement.toStringUtf8());
+        }
+    }
+
+    public Block extractAll(Slice source, long groupIndex)
+    {
+        Matcher matcher = re2jPattern.matcher(source);
+        int group = checkedCast(groupIndex);
+        validateGroup(group, matcher.groupCount());
+
+        BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(new BlockBuilderStatus(), 32);
+        while (true) {
+            if (!matcher.find()) {
+                break;
+            }
+
+            Slice searchedGroup = matcher.group(group);
+            if (searchedGroup == null) {
+                blockBuilder.appendNull();
+                continue;
+            }
+            VARCHAR.writeSlice(blockBuilder, searchedGroup);
+        }
+        return blockBuilder.build();
+    }
+
+    public Slice extract(Slice source, long groupIndex)
+    {
+        Matcher matcher = re2jPattern.matcher(source);
+        int group = checkedCast(groupIndex);
+        validateGroup(group, matcher.groupCount());
+
+        if (!matcher.find()) {
+            return null;
+        }
+
+        return matcher.group(group);
+    }
+
+    public Block split(Slice source)
+    {
+        Matcher matcher = re2jPattern.matcher(source);
+        BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(new BlockBuilderStatus(), 32);
+
+        int lastEnd = 0;
+        while (matcher.find()) {
+            Slice slice = source.slice(lastEnd, matcher.start() - lastEnd);
+            lastEnd = matcher.end();
+            VARCHAR.writeSlice(blockBuilder, slice);
+        }
+
+        VARCHAR.writeSlice(blockBuilder, source.slice(lastEnd, source.length() - lastEnd));
+        return blockBuilder.build();
+    }
+
+    private void validateGroup(int group, int groupCount)
+    {
+        if (group < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Group cannot be negative");
+        }
+        if (group > groupCount) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Pattern has %d groups. Cannot access group %d", groupCount, group));
+        }
+    }
+
+    private static class RE2JEventsListener
+            implements Options.EventsListener
+    {
+        @Override
+        public void fallbackToNFA()
+        {
+            log.info("Fallback to NFA");
```

Comment:
Updated message to be more meaningful.

RE2J will fallback to NFA after `dfaRetries` retries. One it fallbacks, then it will continue to use NFA for the rest of the query. The idea is that there might be few outlier rows that will cause DFA to generate too many states. In such case, NFA will be used for such row and DFA state machine is reset.